runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -154,6 +149,7 @@ public int getSizeEstimate() {
      * Return the payload.
      */
     public LogData(ByteBuf buf) {
+        lastKnownSize = buf.readableBytes();
```

Comment:
Serialized is usually more compact than in-memory representation, not will be sufficient.

The size is used by the server cache only. Since the logunit doesn't deserialize the payload, I think a better estimation function would be to use the byte array size + sizeOf(metadata size).

sizeOf can be slow and consume a lot of memory, but since the metadata structure is very flat, I expect this function to have very low overhead and it will produce an accurate estimation of in memory size.

Reply:
Can we write a benchmark for `sizeOf`? Because you can assume that the sizeOf operation is fast enough but in reality, it would be deadly slow. 
We have benchmarks module, specially designed for that.
This can be used as an example https://github.com/CorfuDB/corfudb-tests/blob/master/benchmarks/src/jmh/java/org/corfudb/benchmarks/util/DataGeneratorBenchmark.java

Reply:
@Maithem Is it worth to use sizeOf? Because metadata is not the only one that has difference between in-memory representation and serialized size. LogUnit caches LogData, even though we calculate the exact in-memory size of metadata plus the payload size, LogData object itself has overheads, which means we can never have a precise size, unless use sizeOf(LogData), which would be more expensive. So does it worth to calculate the accurate in-memory size of metadata?

Reply:
So, I have some numbers.
I have created a LogData object:
```
ByteBuf b = Unpooled.buffer();
            byte[] streamEntry = DataGenerator.generateDataString(1024).getBytes();
            Serializers.CORFU.serialize(streamEntry, b);
            LogData ld = new LogData(DataType.DATA, b);
            ld.setCheckpointedStreamId(UUID.randomUUID());
            ld.setClientId(UUID.randomUUID());
            ld.setCheckpointedStreamStartLogAddress(321L);
            ld.setGlobalAddress(45678L);
            Map<UUID, Long> bm = new HashMap<>();
            for (int i = 0; i < 100; i++) {
                bm.put(UUID.randomUUID(), rnd.nextLong());
            }

            ld.setBackpointerMap(bm);
```
I have compared serialized size vs in-memory:
`Serialized: 4096, sizeof: 11272`
So, 3x difference.

Regarding SizeOf:
The more you have records in the `metadataMap` the slower sizeOf operation.
Without any records in `metadataMap` the performance is about 300k operations on my laptop.
Having just 100 records in the `BackpointerMap<UUID, Long>` makes it much slower ~ 20k op/s

So, both ways (measuring serialized size vs sizeOf) are not good enough.
But measuring the serialized size gives us at least something.

Reply:
And as you can see using: `sizeOf(metadataMap)` is the worst idea

Reply:
Instead of use useOf(a Map), should use sizeOf(logdata). Not sure it will help. Some benchmark may help.

Reply:
Also, it is possible that the overhead is static : (11272 - 4096)? Will the overhead change for different payload objects? If the overhead is static, we can calculate the serialized size + the static overhead.

Reply:
I would suggest to do some benchmark for calculate metadata and logdata:
comparing the total serialized data size and the size we get with sizeOf api.
Is it 40% or 80%.

Reply:
yes, it should be possible to test that using the heapdump we get with the new changes.  We will have lastKnownSize(serialized bytebuf size) in the dump and also the actual logData object with the in-mem size.  

Let us know if you have any other/better ideas.